lib/spack/spack/variant.py

```diff
@@ -35,36 +37,77 @@
 special_variant_values = [None, "none", "*"]
 
 
+class ValidValue:
+    def __init__(self, values):
+        self._values = values
+
+    def __call__(self, v):
+        return v in self._values
+
+    def __eq__(self, other):
+        return self._values == other._values
+
+    def __str__(self):
+        return f"<valid if in {self._values}>"
+
+
+always_true = lambda x: True
+
+
+def variant_class_for_type(variant_type: str) -> type:
+    """Associates string representation of variant type with classes."""
+    variant_class = {
+        "multi": MultiValuedVariant,
+        "bool": BoolValuedVariant,
+        "single": SingleValuedVariant,
+    }.get(variant_type)
+
+    if variant_class is None:
+        raise ValueError(f"Invalid variant type: {variant_type}")
+
+    return variant_class
+
+
+def make_variant(variant_type: str, name: str, value: Any) -> "AbstractVariant":
+    """Make concrete variant instance given a type, name, and value."""
+    cls = variant_class_for_type(variant_type)
+    return cls(name, value)
+
+
 class Variant:
-    """Represents a variant in a package, as declared in the
-    variant directive.
+    """Represents a variant definition, created by the ``variant()`` directive.
+
+    There can be multiple definitions of the same variant, and they are given precedence
+    by order of appearance in the package. Later definitions have higher precedence.
+    Similarly, definitions in derived classes have higher precedence than those in their
+    superclasses.
+
     """
 
     def __init__(
         self,
-        name,
-        default,
-        description,
-        values=(True, False),
-        multi=False,
-        validator=None,
-        sticky=False,
+        name: str,
+        default: Any,
+        description: str,
+        values: Union[Sequence, Callable] = (True, False),
+        multi: bool = False,
+        validator: Optional[Callable] = None,
+        sticky: bool = False,
+        precedence: int = 0,
```

Comment:
Not for this PR, but doing:
```suggestion
        name: str,
        *,
        default: Any,
        description: str,
        values: Union[Sequence, Callable] = (True, False),
        multi: bool = False,
        validator: Optional[Callable] = None,
        sticky: bool = False,
        precedence: int = 0,
```
would help reading code at the caller site.

Reply:
this one was easy. I just did it -- none of the calls failed to do this so it worked fine.